pxtlib/tutorial.ts

```diff
@@ -563,4 +527,37 @@ ${code}
         }
         return res;
     }
+
+    export async function getTutorialHighlightedBlocks(tutorial: TutorialOptions): Promise<pxt.Map<pxt.Map<number>> | undefined> {
+        const db = await pxt.BrowserUtils.tutorialInfoDbAsync();
+        const entry = await db.getAsync(tutorial.tutorial, tutorial.tutorialCode);
+        return entry?.highlightBlocks;
+    }
+
+    export function getTutorialStepHash(tutorial: TutorialOptions): string {
+        const { tutorialStepInfo, tutorialStep } = tutorial;
+        const body = tutorialStepInfo[tutorialStep].hintContentMd;
+        const codeSnippets = getBlockSnippetCode(body);
+        return pxt.BrowserUtils.getTutorialCodeHash(codeSnippets);
+    }
+
+    /** TODO: if this gets exported, we probably want to consider generalizing to 'parseMarkdownSnippets'
+     * that returns { snippetName: string[] }; e.g.
+     * { "blocks": ["block snippet 1", "block snippet 2"], "package": ["deplist 1"] }
+     * need to cover getMetadataRegex + stripHiddenSnippets types, maybe more if any happen to be around.
+     **/
+    function getBlockSnippetCode(mdSnippet: string): string[] {
+        let hintCode: string[] = [];
+
+        mdSnippet?.replace(/((?!.)\s)+/g, "\n")
+            .replace(
+                /``` *(block|blocks)\s*\n([\s\S]*?)\n```/gim,
+                function (m0, m1, m2) {
+                    hintCode.push(`{\n${m2}\n}`);
```

Comment:
minor diff, we were only accounting for first snippet in a hint, probably affects no one but probably good to confirm. We might also want to include package keys in future to avoid packages changing but that's probably fine